lib/core/src/Cardano/Wallet/DB/Store/Wallets/Store.hs

```diff
@@ -17,109 +19,155 @@ Implementation of a store for 'TxWalletsHistory'
+        (SqlPersistT IO)
+        (DeltaMap W.WalletId DeltaWalletsMetaWithSubmissions)
+mkStoreWalletsMetaWithSubmissions =
+    Store
+    { loadS = load
+    , writeS = write
+    , updateS = update
+    }
+  where
+    write reset = forM_ (Map.assocs reset) $ \(wid, ms) ->
+        writeS (mkStoreWalletMetaWithSubmissions wid) ms
+    update :: Map W.WalletId (TxMetaHistory, TxLocalSubmissionHistory)
+        -> DeltaMap W.WalletId DeltaWalletsMetaWithSubmissions
+        -> SqlPersistT IO ()
+    update _ (Insert wid ms) = do
+        writeS (mkStoreWalletMetaWithSubmissions wid) ms
     update _ (Delete wid) = do
         deleteWhere [TxMetaWalletId ==. wid ]
-    update _ (Adjust wid da) =
-        updateS (mkStoreMetaTransactions wid) undefined da
-    load :: SqlPersistT
-            IO
-            (Either SomeException (Map.Map W.WalletId TxMetaHistory))
-    load = do
-        wids <- nub . fmap (view #txMetaWalletId . entityVal)
+        deleteWhere [LocalTxSubmissionWalletId ==. wid ]
+    update old (Adjust wid xda) =
+        updateS (mkStoreWalletMetaWithSubmissions wid)
+            (Map.findWithDefault mempty wid old) xda
+    load = runExceptT $ do
+        wids <- lift $ fmap (view #txMetaWalletId . entityVal)
             <$> selectList @TxMeta [] []
-        runExceptT $ do
-            xs <- forM wids $ ExceptT . loadS . mkStoreMetaTransactions
-            pure $ Map.fromList $ zip wids xs
+        subsWids <- lift $ fmap (view #localTxSubmissionWalletId . entityVal)
+            <$> selectList @LocalTxSubmission [] []
+        fmap Map.fromList
+            $ forM (nub $ wids <> subsWids) $ \wid -> (wid,)
+                <$> ExceptT (loadS $ mkStoreWalletMetaWithSubmissions wid)
 
--- | Store for 'DeltaTxWalletsHistory'.
-mkStoreTxWalletsHistory
-    :: Store (SqlPersistT IO) DeltaTxWalletsHistory
+
+mkStoreTxWalletsHistory :: Store (SqlPersistT IO) DeltaTxWalletsHistory
 mkStoreTxWalletsHistory =
     Store
     { loadS =
-            liftA2 (,) <$> loadS mkStoreTransactions <*> loadS mkStoreWalletsMeta
+          liftA2 (,)
+            <$> loadS mkStoreTransactions
+            <*> loadS mkStoreWalletsMetaWithSubmissions
     , writeS = \(txHistory,txMetaHistory) -> do
-            writeS mkStoreTransactions txHistory
-            writeS mkStoreWalletsMeta txMetaHistory
+          writeS mkStoreTransactions txHistory
+          writeS mkStoreWalletsMetaWithSubmissions txMetaHistory
     , updateS = \(txh@(TxHistoryF mtxh),mtxmh) -> \case
-            ExpandTxWalletsHistory wid cs -> do
-                updateS mkStoreTransactions txh
-                    $ TxStore.Append
-                    $ mkTxHistory
-                    $ fst <$> cs
-                -- see also Store.Wallets.Model.garbageCollectEmptyWallets
-                updateS mkStoreWalletsMeta mtxmh
-                    $ case Map.lookup wid mtxmh of
-                        Nothing ->
-                            Insert wid $ mkTxMetaHistory wid cs
```

Comment:
Ah, this code also relates to my comment on `update old (Adjust wid xda)` — I think making the case distinction here and using `Insert` is something that we have to keep doing. Again, our hands are tied by the `instance Delta (DeltaMap key da)`.

(It was worth a try. 😂 )

Reply:
I think I got it, let's see